src/raftstore/store/fsm/router.rs

```diff
@@ -0,0 +1,627 @@
+    /// no reference to mailbox after calling this method.
+    fn take_mailbox(&mut self) -> Option<BasicMailbox<Self>>;
+}
+
+/// A basic mailbox.
+///
+/// Every mailbox should have one and only one an owner, who will receives all
+/// messages sent to this mailbox.
+///
+/// When a message is sent to a mailbox, its owner will be checked whether it's
+/// idle. An idle owner will be scheduled via `FsmScheduler` immediately, which
+/// will drive the fsm to poll for messages.
+pub struct BasicMailbox<N: Fsm> {
+    sender: mpsc::LooseBoundedSender<N::Message>,
+    state: Arc<State<N>>,
+}
+
+impl<N: Fsm> BasicMailbox<N> {
+    #[inline]
+    fn new(sender: mpsc::LooseBoundedSender<N::Message>, fsm: Box<N>) -> BasicMailbox<N> {
+        BasicMailbox {
+            sender,
+            state: Arc::new(State {
+                status: AtomicUsize::new(NOTIFYSTATE_IDLE),
+                data: AtomicPtr::new(Box::into_raw(fsm)),
+            }),
+        }
+    }
+
+    /// Take the owner if it's IDLE.
+    fn maybe_catch_fsm(&self) -> Option<Box<N>> {
+        match self
+            .state
+            .status
+            .swap(NOTIFYSTATE_NOTIFIED, Ordering::AcqRel)
+        {
+            NOTIFYSTATE_NOTIFIED => return None,
+            NOTIFYSTATE_IDLE => {}
+            _ => return None,
+        }
+
+        let p = self.state.data.swap(ptr::null_mut(), Ordering::AcqRel);
+        if !p.is_null() {
+            Some(unsafe { Box::from_raw(p) })
+        } else {
+            panic!("inconsistent status and data, something should be wrong.");
+        }
+    }
+
+    /// Notify owner via a `FsmScheduler`.
+    #[inline]
+    fn notify<S: FsmScheduler<Fsm = N>>(&self, scheduler: &S) {
+        match self.maybe_catch_fsm() {
+            None => {}
+            Some(mut n) => {
+                n.set_mailbox(Cow::Borrowed(self));
+                scheduler.schedule(n);
+            }
+        }
+    }
+
+    /// Put the owner back to the state.
+    ///
+    /// It's not required that all messages should be consumed before
+    /// releasing a fsm. However, a fsm is guaranteed to be notify only
+    /// when new messages arrives after it's released.
+    #[inline]
+    fn release(&self, fsm: Box<N>) {
+        let previous = self.state.data.swap(Box::into_raw(fsm), Ordering::AcqRel);
+        let mut previous_status = NOTIFYSTATE_NOTIFIED;
+        if previous.is_null() {
+            previous_status = self.state.status.compare_and_swap(
+                NOTIFYSTATE_NOTIFIED,
+                NOTIFYSTATE_IDLE,
+                Ordering::AcqRel,
+            );
+            match previous_status {
+                NOTIFYSTATE_NOTIFIED => return,
+                NOTIFYSTATE_DROP => {
+                    let ptr = self.state.data.swap(ptr::null_mut(), Ordering::AcqRel);
+                    unsafe { Box::from_raw(ptr) };
+                    return;
+                }
+                _ => {}
```

Comment:
Is is reachable?

Reply:
It should be unreachable, that's why L161 panics.